p4include/psa.p4

```diff
@@ -0,0 +1,672 @@
+ *   the packet to drop, and no clone will be created (following the
+ *   analogy with intrinsic metadata bit fields, drop bits are
+ *   processed before clone bits are looked up).
+ *
+ * - resubmit
+ * - recirculate
+ * \TODO: finalize the semantics of calling multiple of the PRE APIs
+ *
+ */
+extern PacketReplicationEngine {
+
+  // PacketReplicationEngine(); /// No constructor. PRE is instantiated
+                                /// by the architecture.
+
+  /// Unicast operation: sends packet to port
+  ///
+  /// Targets may implement this operation by setting the appropriate
+  /// intrinsic metadata or through some other mechanism of
+  /// configuring the PRE.
+  ///
+  /// @param port The output port. If the port is PORT_CPU the packet
+  ///             will be sent to CPU
+  void send_to_port (in PortId_t port);
+
+  /// Multicast operation: sends packet to a multicast group or a port
+  ///
+  /// Targets may implement this operation by setting the appropriate
+  /// intrinsic metadata or through some other mechanism of
+  /// configuring the PRE.
+  ///
+  /// @param multicast_group Multicast group id. The control plane
+  ///        must program the multicast groups through a separate
+  ///        mechanism.
+  void multicast (in MulticastGroup_t multicast_group);
+
+  /// Drop operation: do not forward the packet
+  ///
+  /// The PSA implements drop as an operation in the PRE. While the
+  /// drop operation can be invoked anywhere in the ingress pipeline,
+  /// the semantics supported by the PSA is that the drop will be at
+  /// the end of the pipeline (ingress or egress).
+  void drop      ();
+
+  /// Clone operation: create a copy of the packet and send it to the
+  ///                  specified port.
+  ///
+  /// The PSA specifies four types of cloning, with the packet sourced
+  /// from different points in the pipeline and sent back to ingress
+  /// or to the buffering queue in the egress (@see CloneMethod_t).
+  /// \TODO: needs both source and destination data
+  ///
+  /// @param clone_method  The type of cloning.
+  /// @param port          The port to send the cloned packet to.
+  ///
+  /// Note: based on discussions with Antonin, we can ignore session now
+  void clone     (in CloneMethod_t clone_method, in PortId_t port);
+
+  /// Clone operation: create a copy of the packet with additional
+  ///                  data and send it to the specified port.
+  ///
+  /// The PSA specifies four types of cloning, with the packet sourced
+  /// from different points in the pipeline and sent back to ingress
+  /// or to the buffering queue in the egress (@see CloneMethod_t).
+  /// \TODO: needs both source and destination data
+  ///
+  /// @param clone_method  The type of cloning.
+  /// @param port          The port to send the cloned packet to.
+  /// @param data additional header data attached to the packet
+  void clone<T>  (in CloneMethod_t clone_method, in PortId_t port, in T data);
```

Comment:
For all these 3 methods (clone(), resubmit() and recirculate()) it is totally not clear what happens to the ```T data```.